### PR TITLE
add user model

### PR DESCRIPTION
**Pull request by rajneeshsharma9**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class User < ApplicationRecord
+
+  REGEXP_EMAIL = /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
```
Comment (sawangupta92) [inline app/models/user.rb]: Make global constant file(default_constant.rb) and move it there.

Comment (sawangupta92) [inline app/models/user.rb]: rename

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  
+  enum role: [:customer, :admin]
```
Comment (sawangupta92) [inline app/models/user.rb]: move this in a constant in this file
[:customer, :admin]

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  has_secure_password
+
+  validates :name, presence: true
+  validates :email, uniqueness: true, format: {
```
Comment (sawangupta92) [inline app/models/user.rb]: what if I enter sawan@vinsol.com and SAWAN@vinsol.com are they different emails?

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    with:    REGEXP_EMAIL,
+    message: 'must be a valid email address.'
```
Comment (sawangupta92) [inline app/models/user.rb]: I18n

File: app/models/user.rb (hunk 3 above)
Comment (sawangupta92) [inline app/models/user.rb]: presence?

File: app/models/user.rb (hunk 2 above)
Comment (sawangupta92) [inline app/models/user.rb]: presence?

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  after_initialize do
+    if self.new_record?
+      self.role ||= :customer
+    end
+  end
+
+end
```
Comment (sawangupta92) [inline app/models/user.rb]: Please add extra empty line at the end of each file

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  enum role: [:customer, :admin]
+  has_secure_password
+
+  validates :name, presence: true
+  validates :email, uniqueness: true, format: {
+    with:    REGEXP_EMAIL,
+    message: 'must be a valid email address.'
+  }
+
+  after_initialize do
```
Comment (sawangupta92) [inline app/models/user.rb]: This is not a readable way to write callbacks, you should define methods instead.

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    if self.new_record?
+      self.role ||= :customer
```
Comment (sawangupta92) [inline app/models/user.rb]: Is this really needed?

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  has_secure_password
+
+  validates :name, presence: true
+  validates :email, uniqueness: true, format: {
+    with:    REGEXP_EMAIL,
+    message: 'must be a valid email address.'
+  }
+
+  after_initialize do
+    if self.new_record?
```
Comment (sawangupta92) [inline app/models/user.rb]: why is self required here?

File: app/models/user.rb (hunk 3 above)
Comment (sawangupta92) [inline app/models/user.rb]: add allow blank here, as uniqueness always fire one extra query. We can save it in case no email ID was present

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 5 above)
Comment (rajneeshsharma9) [reply app/models/user.rb]: whats the reason for this?

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
+class User < ApplicationRecord
+  
+  ROLES = { customer: 0, admin: 1 }
+  enum role: ROLES
+  has_secure_password
+
+  validates :name, presence: true
+  validates :email, presence: true
```
Comment (sawangupta92) [inline app/models/user.rb]: can be merged in single line

File: app/models/user.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  validates :email, :uniqueness => { :case_sensitive => false }, format: {
```
Comment (sawangupta92) [inline app/models/user.rb]: can we use new hash syntax here?

File: config/initializers/default_constants.rb [hunk 11]
```diff
@@ -0,0 +1 @@
+EMAIL_REGEXP = /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
```
Comment (sawangupta92) [inline config/initializers/default_constants.rb]: empty line

File: config/locales/models/user/en.yml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+en:
+  invalid_email: "must be a valid email address."
```
Comment (sawangupta92) [inline config/locales/models/user/en.yml]: add some scope as well

Comment (sawangupta92) [inline config/locales/models/user/en.yml]: empty line

File: app/models/user.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
+class User < ApplicationRecord
+  
+  ROLES = { customer: 0, admin: 1 }
```
Comment (sawangupta92) [inline app/models/user.rb]: freeze

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

**Pull request merged**